File: trunk/BC Math Lib/libbcmath.compare.vb
Partial Public Class libbcmath
	''' <summary>Compare two arbitrary precision numbers</summary>
	''' <param name="n1">The first number (1 if larger)</param>
	''' <param name="n2">The second number (-1 if larger)</param>
	Public Shared Function DoCompare(ByRef n1 As BCNum, ByRef n2 As BCNum) As SByte
		Return DoCompareAdvanced(n1, n2, True, False)
	End Function

	''' <summary>Raw compare of two arbitrary precision numbers</summary>
	''' <param name="n1">The first number (1 if larger)</param>
	''' <param name="n2">The second number (-1 if larger)</param>
	''' <param name="use_sign">Check the signs, only magnitude is checked if false</param>
	''' <param name="ignore_last">Whether or not to ignore the last digit</param>
	Protected Shared Function DoCompareAdvanced(ByRef n1 As BCNum, ByRef n2 As BCNum, ByVal use_sign As Boolean, ByVal ignore_last As Boolean) As SByte
		Dim n1ptr As Integer = 0, n2ptr As Integer = 0, count As Integer = 0

		' First, compare signs.
		If use_sign And (n1.sign <> n2.sign) Then
			Return CSByte(If(n1.sign = PLUS, 1, -1)) ' POSITIVE n1 > NEGATIVE n2 = 1
		End If ' the signs must be the same now!

		' Now compare the magnitude.
		If n1.length > n2.length Then
			Return CSByte(If((Not use_sign) Or n1.sign = PLUS, 1, -1))
		ElseIf n1.length < n2.length Then
			Return CSByte(If((Not use_sign) Or n1.sign = PLUS, -1, 1)) ' same sign
		End If

		' If we get here, they have the same number of integer digits.
		' Check the integer part and the equal length part of the fraction.
		count = n1.length + Math.Min(n1.scale, n2.scale)
		While (count > 0) And (n1(n1ptr) = n2(n2ptr))
			n1ptr += 1
			n2ptr += 1
			count -= 1
		End While

		If ignore_last And (count = 1) And (n1.scale = n2.scale) Then Return 0

		If count <> 0 Then
			If n1(n1ptr) > n2(n2ptr) Then : Return CSByte(If((Not use_sign) Or n1.sign = PLUS, 1, -1))
			Else : Return CSByte(If((Not use_sign) Or n1.sign = PLUS, -1, 1))
			End If
		End If

		' They are equal up to the last part of the equal part of the fraction.
		If n1.scale > n2.scale Then
			For count = (n1.scale - n2.scale) To 1 Step -1
				If (n1(n1ptr) <> 0) Then
					Return CSByte(If((Not use_sign) Or n1.sign = PLUS, 1, -1))
				End If
				n1ptr += 1
			Next
		ElseIf n1.scale < n2.scale Then
			For count = (n2.scale - n1.scale) To 1 Step -1
				If (n2(n2ptr) <> 0) Then
					Return CSByte(If((Not use_sign) Or n1.sign = PLUS, -1, 1))
				End If
				n2ptr += 1
			Next
		End If

		' They must be equal!
		Return 0
	End Function
End Class

File: tags/alpha4-fail/libbcmath/libbcmath.compare.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
public partial class libbcmath
{
	/// <summary>Compare two arbitrary precision numbers</summary>
	/// <param name="n1">The first number (1 if larger)</param>
	/// <param name="n2">The second number (-1 if larger)</param>
	public static sbyte DoCompare(BCNum n1, BCNum n2)
	{
		return DoCompareAdvanced(n1, n2, true, false);
	}

	/// <summary>Raw compare of two arbitrary precision numbers</summary>
	/// <param name="n1">The first number (1 if larger)</param>
	/// <param name="n2">The second number (-1 if larger)</param>
	/// <param name="use_sign">Check the signs, only magnitude is checked if false</param>
	/// <param name="ignore_last">Whether or not to ignore the last digit</param>
	private static sbyte DoCompareAdvanced(BCNum n1, BCNum n2, bool use_sign, bool ignore_last)
	{
		int n1ptr = 0;
		int n2ptr = 0;
		int count = 0;

		// First, compare signs.
		if (use_sign & (n1.sign != n2.sign)) {
			return Convert.ToSByte(n1.sign == PLUS ? 1 : -1);
			// POSITIVE n1 > NEGATIVE n2 = 1
		}
		// the signs must be the same now!

		// Now compare the magnitude.
		if (n1.length > n2.length) {
			return Convert.ToSByte((!use_sign) | n1.sign == PLUS ? 1 : -1);
		} else if (n1.length < n2.length) {
			return Convert.ToSByte((!use_sign) | n1.sign == PLUS ? -1 : 1);
			// same sign
		}

		// If we get here, they have the same number of integer digits.
		// Check the integer part and the equal length part of the fraction.
		count = n1.length + Math.Min(n1.scale, n2.scale);
		while ((count > 0) & (n1[n1ptr] == n2[n2ptr])) {
			n1ptr += 1;
			n2ptr += 1;
			count -= 1;
		}

		if (ignore_last & (count == 1) & (n1.scale == n2.scale))
			return 0;

		if (count != 0) {
			if (n1[n1ptr] > n2[n2ptr]) {
				return Convert.ToSByte((!use_sign) | n1.sign == PLUS ? 1 : -1);
			} else {
				return Convert.ToSByte((!use_sign) | n1.sign == PLUS ? -1 : 1);
			}
		}

		// They are equal up to the last part of the equal part of the fraction.
		if (n1.scale > n2.scale) {
			for (count = (n1.scale - n2.scale); count >= 1; count += -1) {
				if ((n1[n1ptr] != 0)) {
					return Convert.ToSByte((!use_sign) | n1.sign == PLUS ? 1 : -1);
				}
				n1ptr += 1;
			}
		} else if (n1.scale < n2.scale) {
			for (count = (n2.scale - n1.scale); count >= 1; count += -1) {
				if ((n2[n2ptr] != 0)) {
					return Convert.ToSByte((!use_sign) | n1.sign == PLUS ? -1 : 1);
				}
				n2ptr += 1;
			}
		}

		// They must be equal!
		return 0;
	}
}